src/main/java/org/traccar/protocol/LaipacProtocolDecoder.java

```diff
@@ -81,81 +100,166 @@ private String decodeAlarm(String event) {
+        position.setValid(upperCaseStatus.equals("A"));
         position.set(Position.KEY_STATUS, status);
 
+        // Position
         position.setLatitude(parser.nextCoordinate());
         position.setLongitude(parser.nextCoordinate());
+
+        // Speed
         position.setSpeed(parser.nextDouble(0));
+
+        // Course
         position.setCourse(parser.nextDouble(0));
 
+        // Date
         dateBuilder.setDateReverse(parser.nextInt(0), parser.nextInt(0), parser.nextInt(0));
         position.setTime(dateBuilder.getDate());
 
+        // Alarm / Event
         String event = parser.next();
         position.set(Position.KEY_ALARM, decodeAlarm(event));
-        position.set(Position.KEY_EVENT, event);
+        position.set(Position.KEY_EVENT, decodeEvent(event, position));
+
+        // Battery
         position.set(Position.KEY_BATTERY, Double.parseDouble(parser.next().replaceAll("\\.", "")) * 0.001);
+
+        // Odometer
         position.set(Position.KEY_ODOMETER, parser.nextDouble());
+
+        // GPS status
         position.set(Position.KEY_GPS, parser.nextInt());
+
+        // ADC1 / ADC2
         position.set(Position.PREFIX_ADC + 1, parser.nextDouble() * 0.001);
         position.set(Position.PREFIX_ADC + 2, parser.nextDouble() * 0.001);
 
-        Integer lac = parser.nextHexInt();
-        Integer cid = parser.nextHexInt();
-        Integer mcc = parser.nextInt();
-        Integer mnc = parser.nextInt();
-        if (lac != null && cid != null && mcc != null && mnc != null) {
+        // LAC, CID, MCC, MNC
+        Integer laccid = parser.nextHexInt();
+        Integer mccmnc = parser.nextInt();
+        if (laccid != null && laccid != 0 && mccmnc != null && mccmnc != 0) {
+            Integer lac = (laccid >> 16) & 0xFFFF;
+            Integer cid = laccid & 0xFFFF;
+            Integer mcc = mccmnc / 10000;
+            Integer mnc = mccmnc % 100;
             position.setNetwork(new Network(CellTower.from(mcc, mnc, lac, cid)));
         }
 
-        String checksum = parser.next();
+        // Skip remaining parameters
+        String unused = parser.next();
+
+        // Checksum
+        String checksumStr = parser.next();
+        if (checksum != Integer.parseInt(checksumStr, 16)) {
+            return null;
+        }
 
         if (channel != null) {
+
+            if (Character.isLowerCase(status.charAt(0))) {
+                String ack = "$EAVACK," + event + "," + checksumStr;
+                ack += Checksum.nmea(ack) + "\r\n";
+                channel.writeAndFlush(new NetworkMessage(ack, remoteAddress));
+            }
+
+            String response = "";
+            String devicePassword = getDevicePassword(deviceSession);
+
             if (event.equals("3")) {
-                channel.writeAndFlush(new NetworkMessage("$AVCFG,00000000,d*31\r\n", remoteAddress));
+                response = "$AVCFG," + devicePassword + ",d";
```

Comment:
I would probably make a separate method instead.

Reply:
Good. I'll fix it.